### PR TITLE
RWAHS-433 Load cleaned `LocationHomes` data

### PR DESCRIPTION
**Pull request by kehh**
- Update refinery parameters to support importing this data as a hierarchy
- Load the cleaned data into a table so we can look up cleaned multi column values

### CODE REVIEW

File: importScripts/refineryParameters/UsualLocationLowest.ca_storage_locations.usual.storageLocationSplitter.json [hunk 1]
```diff
@@ -0,0 +1,47 @@
+{
+    "storageLocationType": "^UsualLocationType",
```
Comment (kehh) [inline importScripts/refineryParameters/UsualLocationLowest.ca_storage_locations.usual.storageLocationSplitter.json:2]: previously `UsualLocation` was a single value, not hierarchical columns. We need a more complicated `storageLocationSplitter` to split out the values.

File: importScripts/refineryParameters/UsualLocationLowest.ca_storage_locations.usual.storageLocationSplitter.json [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    "parents": [
+        {
+            "name": "^UsualLocationShelf",
+            "type": "shelf",
+            "attributes": {
+                "is_enabled": 1
+            },
+            "rules": [
+                {
+                    "trigger": "^UsualLocationType = \"shelf\"",
```
Comment (kehh) [inline importScripts/refineryParameters/UsualLocationLowest.ca_storage_locations.usual.storageLocationSplitter.json:20]: Skip this level as the base level had a `^UsualLocationType` of `shelf`

File: importScripts/refineryParameters/UsualLocationLowest.ca_storage_locations.usual.storageLocationSplitter.json [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        },
+        {
+            "name": "^UsualLocationRoom",
+            "type": "room",
+            "attributes": {
+                "is_enabled": 1
+            },
+            "rules": [
+                {
+                    "trigger": "^UsualLocationType = \"room\"",
```
Comment (kehh) [inline importScripts/refineryParameters/UsualLocationLowest.ca_storage_locations.usual.storageLocationSplitter.json:37]: "

File: importScripts/sampleRecords/additionalAdministration.browse.bundles.json [hunk 4]
```diff
@@ -12,6 +12,8 @@
         "ca_objects.OtherNumber": {
             "delimiter": ";"
         },
-        "ca_storage_locations": {}
+        "usualLocationHierarchy": {
```
Comment (kehh) [inline importScripts/sampleRecords/additionalAdministration.browse.bundles.json:15]: output this data as a hierarchy

File: importScripts/sampleRecords/additionalAdministration.destination.json [hunk 5]
```diff
@@ -9,8 +9,7 @@
       "ca_objects.AccessionByContainer.AccessionBy":"Curator",
       "ca_objects.is_deaccessioned":"0",
       "ca_objects.LastEditDate":"30 July 2015",
-      "ca_objects.LastEditBy":"PENDS",
-      "ca_storage_locations":"Artworks Store Room C2"
```
Comment (kehh) [inline importScripts/sampleRecords/additionalAdministration.destination.json:13]: just investigating where these values went :(

File: importScripts/sql/transformations/ObjectsExportView.sql [hunk 6]
```diff
@@ -125,6 +125,28 @@ CREATE OR REPLACE VIEW ObjectsExport AS
         NULLIF(TRIM(o.TreatmentCompleteYesNo), '')                AS TreatmentCompleteYesNo,
         NULLIF(TRIM(o.TreatmentPriority), '')                     AS TreatmentPriority,
         NULLIF(TRIM(o.UsualLocation), '')                         AS UsualLocation,
+        lhc.Room                                                  AS UsualLocationRoom,
+        lhc.Shelf                                                 AS UsualLocationShelf,
+        lhc.Box                                                   AS UsualLocationBox,
+        IF(
+            o.ItemType IN ('Photograph', 'Memorials'),
```
Comment (kehh) [inline importScripts/sql/transformations/ObjectsExportView.sql:132]: Memorials are stored in `ca_places`, Photographs linked to a list item.

File: importScripts/sampleRecords/additionalAdministration.destination.json [hunk 7]
```diff
@@ -10,7 +10,7 @@
       "ca_objects.is_deaccessioned":"0",
       "ca_objects.LastEditDate":"30 July 2015",
       "ca_objects.LastEditBy":"PENDS",
-      "ca_storage_locations":"Artworks Store Room C2"
+      "usualLocationHierarchy":" (id:1, ); Artworks Store Room C2 (id:2, room)"
```
Comment (kehh) [inline importScripts/sampleRecords/additionalAdministration.destination.json]: now it's a hierarchy, the root is the top level (id:1)

File: importScripts/sampleRecords/additionalAdministration.destination.json [hunk 8]
```diff
@@ -10,7 +10,7 @@
       "ca_objects.is_deaccessioned":"0",
       "ca_objects.LastEditDate":"30 July 2015",
       "ca_objects.LastEditBy":"PENDS",
-      "ca_storage_locations":"Artworks Store Room C2"
+      "usualLocationHierarchy":" (id:1, ) => Artworks Store Room C2 (id:2, room)"
```
Comment (kehh) [inline importScripts/sampleRecords/additionalAdministration.destination.json:13]: This is a hierarchy now. The root has id:1 and no type. The rest of the values have a type.

File: importScripts/sampleRecords/additionalAdministration.destination.json [hunk 9]
```diff
@@ -21,23 +21,56 @@
...
-      "id":"4",
+      "object_id":"6",
+      "id":"6",
+      "idno":"C1985.121a",
+      "display_label":"DRESS, FAWN TUSSORE SILK.",
+      "ca_objects.AccessionByContainer.AccessionBy":"UEB",
+      "ca_objects.is_deaccessioned":"0",
+      "ca_objects.LastEditDate":"21 February 2013",
+      "ca_objects.LastEditBy":"ValH",
+      "usualLocationHierarchy":" (id:1, ) => Costume Room (id:9, room) => Box 099 (id:10, box)"
```
Comment (kehh) [inline importScripts/sampleRecords/additionalAdministration.destination.json:69]: This `Costume Room` has the same id as the one for `C2003.1e` below

**Comment by kehh** [comment]
OK now I've really finished making comments.

**Comment by leftclickben** [comment]
@kehh Did you mean to close and delete this?

**Comment by kehh** [comment]
No I don't think so.

**Comment by kehh** [comment]
Will try to find the deleted branch.

**Comment by kehh** [comment]
So I push deleted the wrong branch. Luckily I still had the branch locally, and the PR can be reopened.

File: importScripts/sampleRecords/additionalAdministration.destination.json [hunk 10]
```diff
@@ -142,18 +174,18 @@
       "ca_objects.deaccession_type_id":"Item discarded",
       "ca_objects.LastEditDate":"16 February 2011",
       "ca_objects.LastEditBy":"VandB",
-      "ca_storage_locations":"Costume Room Box 225"
+      "usualLocationHierarchy":" (id:1, ) => Costume Room Box 225 (id:13, room)"
```
Comment (leftclickben) [inline importScripts/sampleRecords/additionalAdministration.destination.json:177]: Is this supposed to be a `room`?

File: importScripts/sampleRecords/additionalAdministration.destination.json [hunk 11]
```diff
@@ -142,18 +174,18 @@
...
+      "object_id":"16",
+      "id":"16",
       "idno":"C1995.36",
       "display_label":"FAN, SMALL, BLACK",
       "ca_objects.AccessionByContainer.AccessionBy":"JILLM",
       "ca_objects.is_deaccessioned":"1",
       "ca_objects.LastEditDate":"15 February 2013",
       "ca_objects.LastEditBy":"ValH",
-      "ca_storage_locations":"Deaccessioned"
+      "usualLocationHierarchy":" (id:1, ) => Deaccessioned (id:14, room)"
```
Comment (leftclickben) [inline importScripts/sampleRecords/additionalAdministration.destination.json:188]: Probably a data issue, but is this right?

**Comment by leftclickben** [comment]
Only a couple of questions that are probably just data related.

File: importScripts/sampleRecords/additionalAdministration.destination.json (hunk 10 above)
Comment (kehh) [inline importScripts/sampleRecords/additionalAdministration.destination.json:177]: This comes from the data that they haven't cleaned - the data stored in the Objects table are different from the values in the LocationHomes table which was what the "controlled" vocabulary of storage locations. I gave them the LocationHomes table for cleaning, and after loading that found that a number of records (including this one) no longer had a storage location.

File: importScripts/sql/transformations/ObjectsExportView.sql [hunk 12]
```diff
@@ -125,6 +125,28 @@ CREATE OR REPLACE VIEW ObjectsExport AS
...
+        CASE
+        WHEN lhc.Box IS NOT NULL
+            THEN 'box'
+        WHEN lhc.Shelf IS NOT NULL
+            THEN 'shelf'
+        WHEN lhc.Room IS NOT NULL
+            THEN 'room'
+        WHEN
+            (o.ItemType NOT IN ('Photograph', 'Memorials') AND
+             NULLIF(TRIM(o.UsualLocation), '') IS NOT NULL AND
```
Comment (kehh) [inline importScripts/sql/transformations/ObjectsExportView.sql:145]: This is where I'm defaulting the location value to `UsualLocation` if it's not one of the standardised `LocationHomes`.

File: importScripts/sampleRecords/additionalAdministration.destination.json (hunk 10 above)
Comment (kehh) [inline importScripts/sampleRecords/additionalAdministration.destination.json:177]: I've put a comment where I did this in the query for the source data below.

File: importScripts/sampleRecords/additionalAdministration.destination.json (hunk 11 above)
Comment (kehh) [inline importScripts/sampleRecords/additionalAdministration.destination.json:188]: affirmative

**Pull request merged**